### ISSUE TITLE
Automatic address discovery

### ISSUE THREAD

**Issue by mvaivre** [alephium/desktop-wallet#136]
Address discovery should be triggered in two cases:

1. When restoring a wallet
2. Manually, for cases where a user is using a same account on multiple devices

This should follow the strategy defined in BIP44 and should be implemented on the backend by @tdroxler (following discussions with @polarker). 

Details must be discussed, but this feature will be considered in my future mockups. 
This will ship *after* 1.2.0 (address derivation). Before that, user will need to "rederive" addresses manually.

**Comment by mvaivre** [comment]
@tdroxler This one would improve our UX a lot! Do you think we could increase priority and start looking at it / discussing it? Thanks :)

**Comment by tdroxler** [comment]
Sure, let's have a call soon.

**Comment by tdroxler** [comment]
I just checked which endpoint should be called and I think I'll rather add a new one `/addresses/<address>/active` which return `true/false` if a tx exists for that address. All other endpoint are way to heavy for that task, while we can find a super light sql query to do that check.

**Comment by LeeAlephium** [comment]
May want to even make it take a range instead to reduce the network requests? `/addresses/active?offset=0&range=24` so the typical case of 4 addresses plus 20 unused

**Comment by mvaivre** [comment]
The endpoint mentioned by Thomas is for one address only. It basically simply check if TX were done with that address.

**Comment by tdroxler** [comment]
> May want to even make it take a range instead to reduce the network requests? `/addresses/active?offset=0&range=24` so the typical case of 4 addresses plus 20 unused, and can return a `bool[]`

we can't do this from the backend as we don't have access to the key used to derive next addresses

**Comment by LeeAlephium** [comment]
@tdroxler user can pass a set of addresses then instead as a json array: `/addresses/active` payload: `['0xoeauaoueao', '0xsaotehusoatehuoatheua', ...]`

**Comment by nop33** [comment]
> Otherwise we're going to have an initial 80 requests - 20 requests per group...

That's only true if the first 80 address indexes have 20 addresses per group. It will definitely be more than that, since we are scanning sequentially.

**Comment by nop33** [comment]
I run a script that iterates through address indexes sequentially until it finds 20 addresses per group. I run it 100 times, and the average total addresses that had to be discovered is `109`.

<details><summary>Click here for source code</summary>

```js
const sdk = require("@alephium/sdk");
const wallet = sdk.walletGenerate();

const totals = [];

for (let i = 0; i < 100; i++) {
  const group0addresses = [];
  const group1addresses = [];
  const group2addresses = [];
  const group3addresses = [];
  const addressIndexesToSkip = [];

  while (
    group0addresses.length < 20 ||
    group1addresses.length < 20 ||
    group1addresses.length < 20 ||
    group2addresses.length < 20
  ) {
    const newAddress = sdk.deriveNewAddressData(wallet.seed, undefined, undefined, addressIndexesToSkip);
    const group = sdk.groupOfAddress(newAddress.address);
    if (group === 0) {
      group0addresses.push(newAddress);
    } else if (group === 1) {
      group1addresses.push(newAddress);
    } else if (group === 2) {
      group2addresses.push(newAddress);
    } else if (group === 3) {
      group3addresses.push(newAddress);
    }
    addressIndexesToSkip.push(newAddress.addressIndex);
    console.log(newAddress.address, group);
  }

  const total = group0addresses.length + group1addresses.length + group2addresses.length + group3addresses.length;

  console.log(`Group 0: ${group0addresses.length}`);
  console.log(`Group 1: ${group1addresses.length}`);
  console.log(`Group 2: ${group2addresses.length}`);
  console.log(`Group 3: ${group3addresses.length}`);
  console.log(`Total: ${total}`);

  totals.push(total);
}

console.log("Average: ", totals.reduce((a, b) => a + b, 0) / totals.length);

```

</details>

**Comment by LeeAlephium** [comment]
Hahah yeah, I forgot that they're indeterminate so it will definitely be more. Nice experiment man! I love it :D

**Comment by tdroxler** [comment]
@LeeAlephium yes totally doable and better to suffer only once the http layer.
Response will be bool[] with (of course) the same order as the passed addresses

**Comment by tdroxler** [comment]
I implemented the new way [here](https://github.com/alephium/explorer-backend/pull/291)
Path is `POST /addresses-active`, I can't use `/addresses/active` otherwise the http server think `active` is an address

**Comment by tdroxler** [comment]
We should also limit the max number of addresses to pass, what's a good number for you? 
As I wrote in the PR:  1000 addresses take around 0.1s, 4000 address 0.5s. 
We clearly don't need that much, maybe 200 is a good limit?

**Comment by nop33** [comment]
200 sounds good to me

**Comment by tdroxler** [comment]
or as cheng proposed: 
>maybe 4 * 20 since the gas limit of address discovery is 20 by default?
so 80 it could also automatically scale when we increase group number?

**Comment by nop33** [comment]
Alright, in this case, we'll need min 2 requests, which is fine

**Comment by tdroxler** [comment]
Egal for me,, what's the ideal number for you? based on group number?

**Comment by tdroxler** [comment]
nice explanation thx!

Then let's do `groupNum * 20` as 20 is the number given by the BIP, at least it will make sense in few months when we see that code again :sweat_smile:

**Comment by tdroxler** [comment]
new endpoint is available here: 

```
curl -X POST 'https://mainnet-explorer-v18-0-backend.alephium.org/addresses-active' -d '["arst","1DKWCEc9HjhchKoMCMA4J1nbNGNjrAoPFnkcgA1w6h7jh"]'

[false,true]
```

**Comment by LeeAlephium** [comment]
@nop33 is right though, this'll be most likely _always_ 2 requests... I think there's a lot of evidence that 100 is a good number...!

**Comment by polarker** [comment]
As we are not generating new addresses for each new transaction by default, 20 is a more than enough gap for address discovery.

I think I saw projects using around 5 addresses for discovery for Ethereum-like projects. I might not remember it exactly, but it should not be far from it.

**Comment by nop33** [comment]
For what it's worth, `groupNum * 20 * 2` would be a much better alternative imo :) But I'll stop discussing here and focus on building now 😇

**Comment by nop33** [comment]
This was implemented and merged into the `v2.0` branch, it can be closed.
